metadata-ingestion/src/datahub/api/entities/dataset/dataset.py

```diff
@@ -533,14 +917,221 @@ def from_datahub(cls, graph: DataHubGraph, urn: str) -> "Dataset":
+
+                # Replace the fields in the result with the processed ones
+                schema_data["fields"] = processed_fields
+
+        return result
+
     def to_yaml(
         self,
         file: Path,
-    ) -> None:
+    ) -> bool:
+        """
+        Write model to YAML file only if content has changed.
+        Preserves comments and structure of the existing YAML file.
+        Returns True if file was written, False if no changes were detected.
+        """
+        # Create new model data
+        # Create new model data with dict() for Pydantic v1
+        new_data = self.dict(exclude_none=True, exclude_unset=True, by_alias=True)
+
+        # Set up ruamel.yaml for preserving comments
+        yaml_handler = YAML(typ="rt")  # round-trip mode
+        yaml_handler.default_flow_style = False
+        yaml_handler.preserve_quotes = True  # type: ignore[assignment]
+        yaml_handler.indent(mapping=2, sequence=2, offset=0)
+
+        if file.exists():
+            try:
+                # Load existing data with comments preserved
+                with open(file, "r") as fp:
+                    existing_data = yaml_handler.load(fp)
+
+                # Determine if the file contains a list or a single document
+                if isinstance(existing_data, dict):
+                    existing_data = [existing_data]
+                if isinstance(existing_data, list):
+                    # Handle list case
+                    updated = False
+                    identifier = "urn"
+                    model_id = self.urn
+
+                    if model_id is not None:
+                        # Try to find and update existing item
+                        for item in existing_data:
+                            existing_dataset = Dataset(**item)
+                            item_identifier = item.get(identifier, existing_dataset.urn)
+                            if item_identifier == model_id:
+                                # Found the item to update - preserve structure while updating values
+                                updated = True
+                                if (
+                                    existing_dataset.schema_metadata
+                                    and existing_dataset.schema_metadata.file
+                                ):
+                                    # Preserve the existing schema file path
+                                    new_data["schema"]["file"] = (
+                                        existing_dataset.schema_metadata.file
+                                    )
+                                    # Check if the content of the schema file has changed
+                                    with open(
+                                        existing_dataset.schema_metadata.file
+                                    ) as schema_fp:
+                                        schema_fp_content = schema_fp.read()
```

Comment:
as a rule of thumb - when you've got this much indentation, you probably need to refactor something